riscv-page-tables/src/page_table.rs

```diff
@@ -1207,6 +1283,137 @@ impl<T: PagingMode> GuestStagePageTable<T> {
+        let mut inner = self.inner.lock();
+        let mut prev_paddr: Option<SupervisorPageAddr> = None;
+        let mut va = vaddr;
+        while va < end {
+            let pte = inner.get_invalidated_leaf(va)?;
+            let paddr = pte.page_addr();
+            let page_size = pte.level().leaf_page_size();
+
+            // Check the page matches the expected page size (one size smaller than the
+            // requested page size)
+            if page_size != expected_page_size {
+                return Err(Error::InvalidPageSize);
+            }
+
+            // Perform some checks on the physical address
+            if let Some(addr) = &prev_paddr {
+                // Check the current page is contiguous with the previous page
+                if addr
+                    .checked_add_pages_with_size(1, expected_page_size)
+                    .ok_or(Error::AddressOverflow)?
+                    != paddr
+                {
+                    return Err(Error::PageNotContiguous);
+                }
+            } else {
+                // This is the first page, we must ensure it's suitably aligned with the requested
+                // page size.
+                if !paddr.is_aligned(requested_page_size) {
+                    return Err(Error::AddressMisaligned(paddr.bits()));
+                }
+            }
+            prev_paddr = Some(paddr);
+
+            // Run the predicate
+            if !pred(paddr, page_size) {
+                return Err(Error::PredicateFailed);
+            }
+
+            // Move to next page
+            va = Self::check_and_increment_vaddr(va, end, page_size)?;
+        }
+
+        // Promotes the non-leaf PTE and retrieves the now-unused page-table page
+        let paddr = inner.get_invalidated_leaf(vaddr)?.page_addr();
+        let (leaf, free_page) = inner.promote_pte(vaddr.into(), paddr, requested_page_size)?;
+
+        Ok((leaf.page_addr(), leaf.level().leaf_page_size(), free_page))
+    }
+
+    /// Verifies the virtual address range is covering only one invalidated page that matches the
+    /// expected page size. Also checks the page state is either "Blocked" and uniquely owned by
+    /// the TVM, or "BlockedShared" and not owned by the TVM. The page must have been invalidated
+    /// with a TLB version that's older than the current one.
+    pub fn demote_page<F>(
+        &self,
+        vaddr: PageAddr<T::MappedAddressSpace>,
+        requested_page_size: PageSize,
+        free_page: Page<InternalClean>,
+        mut pred: F,
+    ) -> Result<(SupervisorPageAddr, PageSize)>
+    where
+        F: FnMut(SupervisorPageAddr, PageSize) -> bool,
+    {
+        let expected_page_size = requested_page_size
+            .size_up()
+            .ok_or(Error::InvalidPageSize)?;
+        vaddr
+            .checked_add_pages_with_size(1, expected_page_size)
+            .ok_or(Error::AddressOverflow)?;
+        let mut inner = self.inner.lock();
```

Comment:
same comment as above. check if vaddr is aligned first, and then you know it can't overflow.